Store/COSQLiteStorePersistentRootBackingStore.m

```diff
@@ -662,57 +662,78 @@ - (BOOL) deleteRevids: (NSIndexSet *)revids
-     FROM c AS commits
-     LEFT OUTER JOIN (SELECT deltabase, MAX(revid) AS maxkeptrevid FROM c WHERE garbage = 0 GROUP BY deltabase) AS info
-     USING (deltabase)
-     WHERE (commits.revid > info.maxkeptrevid) OR info.maxkeptrevid IS NULL;
+     insert into c values(4,4,3,0);
+     insert into c values(5,4,4,0);
      
-     -- The "OR info.maxkeptrevid IS NULL" part is so that we delete all commits in delta runs where all commits are garbage.
+     insert into c values(6,6,5,0);
+     insert into c values(7,6,6,1); -- marked as garbage, will be selected for deletion
+     insert into c values(8,6,7,0); -- will be rebuit because it's not a full snapshot, but its parent is being deleted
      
+	 
+	 -- Identifies the revisions that need to be rebuilt:
+	 
+	 SELECT revid
+	 FROM c
+	 LEFT OUTER JOIN (SELECT garbage AS parentgarbage, revid AS parentrevid FROM c)
+	 ON (parent = parentrevid)
+	 WHERE garbage = 0 AND parentgarbage = 1 AND deltabase != revid;
+	 
      */
-    
-    // Could be done at a later time
-    [db_ executeUpdate: [NSString stringWithFormat: @"DELETE FROM %@ WHERE revid IN (SELECT commits.revid "
-                         "FROM %@ AS commits "
-                         "LEFT OUTER JOIN (SELECT deltabase, MAX(revid) AS maxkeptrevid FROM %@ WHERE garbage = 0 GROUP BY deltabase) AS info "
-                         "USING (deltabase) "
-                         "WHERE (commits.revid > info.maxkeptrevid) OR info.maxkeptrevid IS NULL)",
-                         [self tableName], [self tableName], [self tableName]]];
-    
-    // TODO: Vacuum here?
-    
+	
+	// Gather the set of revids that need to be rebuilt
+	NSMutableIndexSet *rebuildRevids = [NSMutableIndexSet indexSet];
+	FMResultSet *rs = [db_ executeQuery: [NSString stringWithFormat:
+										  @"SELECT revid "
+										  "FROM %@ "
+										  "LEFT OUTER JOIN (SELECT garbage AS parentgarbage, revid AS parentrevid FROM %@) "
+										  "ON (parent = parentrevid) "
+										  "WHERE garbage = 0 AND parentgarbage = 1 AND deltabase != revid", [self tableName], [self tableName]]];
+	while ([rs next])
+	{
+		[rebuildRevids addIndex: [rs longLongIntForColumnIndex: 0]];
+	}
+	[rs close];
+	
+	// Rebuild each revision that needs it
+	[rebuildRevids enumerateIndexesUsingBlock: ^(NSUInteger revid, BOOL *stop){
+		COItemGraph *graph = [self itemGraphForRevid: revid];
+		
+		// GC unreachable items in graph
+		[graph removeUnreachableItems];
+		
+		NSData *contentsBlob = contentsBLOBWithItemTree(graph);
+		
+		BOOL ok = [db_ executeUpdate: [NSString stringWithFormat: @"UPDATE %@ SET contents = ?, hash = ?, deltabase = ?, bytesInDeltaRun = ? WHERE revid = ?", [self tableName]],
+				   /* contents        */ contentsBlob,
+				   /* hash            */ Sha1Data(contentsBlob),
+				   /* deltabase       */ @(revid),
+				   /* bytesInDeltaRun */ @([contentsBlob length]),
+				   /* revid           */ @(revid)];
+
+		if (!ok)
+		{
+			[db_ rollback];
```

Comment:
The transaction wouldn't be committed unless we get to the -commit call, so the DB wouldn't be corrupted.